### ISSUE TITLE
Regarding TP, FP, FN in confusion matrix and P, R and F1 score in validation results.

### ISSUE THREAD

**Issue by beenajambucha1-design** [ultralytics/ultralytics#24855]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

(1) I have trained my dataset with yolov8n model. Then using its best.pt weights, i tried validation of validation dataset using below code.

<img width="613" height="182" alt="Image" src="https://github.com/user-attachments/assets/df3ad154-9203-44bf-89bb-3fc4aa2afeb5" />

It gave output as below: 

<img width="558" height="126" alt="Image" src="https://github.com/user-attachments/assets/01e28946-51d9-462a-aedc-b4f7f1c8c8d2" />

 It gave me confusion matrix a shown below. 

<img width="494" height="400" alt="Image" src="https://github.com/user-attachments/assets/5a3fb002-5d98-466c-9197-67425678b089" />

As per this confusion matrix, FP for bolls class is 34, FP for blossom class is 11,  FP for split class is 2 and FP for matured class is 17. If we talk about TP, TP for every class is 101, 112, 113, and 120. From this digits, as per equation of Precision = TP/(TP+FP) it should be 0.7481, 0.9105, 0.9826 and 0.8759 respectively for Boll, blossom, split and matured class. Why this discrepancy happens?

(2) Also i used following code to classify only FP bounding boxes from all predicted bounding boxes. 

[CodetofetchFP.docx](https://github.com/user-attachments/files/29040767/CodetofetchFP.docx)
 
And it gave me following results which are again different than above two.
Class: boll (ID: 0)
  True Positives: 101 boxes
  False Positives: 35 boxes
  False Negatives: 16 boxes

Class: blossom (ID: 1)
  True Positives: 114 boxes
  False Positives: 11 boxes
  False Negatives: 7 boxes

Class: split (ID: 2)
  True Positives: 113 boxes
  False Positives: 3 boxes
  False Negatives: 0 boxes

Class: matured (ID: 3)
  True Positives: 118 boxes
  False Positives: 22 boxes
  False Negatives: 11 boxes

Where am i doing mistake?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @beenajambucha1-design, thanks for raising this detailed question about validation metrics and confusion matrix interpretation in Ultralytics 🚀. This is an automated response to help gather the right details, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a custom training ❓ question, please provide as much information as possible, including:
- The exact validation code or command used
- Your model version and `ultralytics` package version
- The saved validation outputs/logs
- The dataset config and class names
- The confidence and IoU settings used during validation and in your FP/TP/FN post-processing script
- A small reproducible sample if possible
- Confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips)

If you believe this may be a 🐛 bug report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it. In particular, since your manually computed TP/FP/FN counts differ from both the confusion matrix and reported validation metrics, a minimal script and sample data would be very helpful 🔍

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Duplicate of #17047

**Comment by Y-T-G** [comment]
Hello.

1. The recall and precision are not based on confusion matrix. You can read about it here:
https://github.com/ultralytics/ultralytics/issues/17047#issuecomment-2424660435

2. Please post your code as a code snippet instead of a `.docx` file.

**Comment by beenajambucha1-design** [comment]
Thank you for clarification. Now my issue is i want to find out all the FP, FN and TP bounding boxes which model have used to calculate P, R and F1 score because i want to analyze those objects which model has declared as FP or FN. Its also ok if its not having specific threshold. But I want to get all those bounding boxes. By applying visualization = True, it only gives data from confusion matrix.

**Comment by beenajambucha1-design** [comment]
I validated the model with confidence threshold of 0.001 which raised my mAP but P value remains same. I also used following code snippet to find out FPs, TPs with confidence threshold of 0.001. But the outputs are same. No change. 

import numpy as np
import torch
# Define IoU threshold and Confidence threshold
iou_threshold = 0.45
confidence_threshold = 0.001 # Added confidence threshold

def xywhn2xyxy(x, w=640, h=640, padw=0, padh=0):
    # Convert nx4 boxes from [x, y, w, h] normalized to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    # x = [x_center, y_center, width, height] normalized
    y = x.copy() * 1.0
    y[0] = w * (x[0] - x[2] / 2) + padw  # x1
    y[1] = h * (x[1] - x[3] / 2) + padh  # y1
    y[2] = w * (x[0] + x[2] / 2) + padw  # x2
    y[3] = h * (x[1] + x[3] / 2) + padh  # y2
    return y.astype(int)

def box_iou(box1: torch.Tensor, box2: torch.Tensor, eps: float = 1e-7) -> torch.Tensor:

  (a1, a2), (b1, b2) = box1.float().unsqueeze(1).chunk(2, 2), box2.float().unsqueeze(0).chunk(2, 2)
  inter = (torch.min(a2, b2) - torch.max(a1, b1)).clamp_(0).prod(2)

    # IoU = inter / (area1 + area2 - inter)
  return inter / ((a2 - a1).prod(2) + (b2 - b1).prod(2) - inter + eps)


classified_detections = {}

for i, img_path in enumerate(selected_images):
    img_height, img_width = results[i].orig_shape # Get original image dimensions

    # Get ground truth annotations for the current image
    gt_boxes_yolo = ground_truth_annotations.get(img_path, [])
    gt_boxes_abs = []
    for gt_anno in gt_boxes_yolo:
        gt_class_id = gt_anno['class_id']
        gt_bbox_norm = np.array(gt_anno['bbox']) # normalized [x_center, y_center, width, height]
        gt_bbox_abs = xywhn2xyxy(gt_bbox_norm, w=img_width, h=img_height) # absolute [x1, y1, x2, y2]
        gt_boxes_abs.append({'class_id': gt_class_id, 'bbox_abs': gt_bbox_abs, 'matched': False})

    # Get model predictions for the current image
    pred_boxes = results[i].boxes.cpu().numpy() # Boxes object from ultralytics results
    preds_abs = []
    # Corrected: Access shape directly from the Boxes object's data attribute or xyxy attribute
    if pred_boxes.xyxy.shape[0] > 0:
        for j in range(pred_boxes.xyxy.shape[0]):
            if pred_boxes.conf[j] >= confidence_threshold: # Filter by confidence threshold
                pred_bbox_xyxy = pred_boxes.xyxy[j] # absolute [x1, y1, x2, y2]
                pred_class_id = int(pred_boxes.cls[j])
                pred_confidence = pred_boxes.conf[j]
                preds_abs.append({'class_id': pred_class_id, 'bbox_abs': pred_bbox_xyxy, 'confidence': pred_confidence, 'matched': False})

    # Initialize lists for this image
    true_positives = []
    false_positives = []
    false_negatives = []

    # Compare ground truth with predictions
    for gt_idx, gt in enumerate(gt_boxes_abs):
        best_iou = 0
        best_pred_idx = -1

        for pred_idx, pred in enumerate(preds_abs):
            if not pred['matched'] and gt['class_id'] == pred['class_id']:
                # Convert numpy arrays to torch tensors and add a batch dimension
                gt_bbox_tensor = torch.from_numpy(gt['bbox_abs']).float().unsqueeze(0)
                pred_bbox_tensor = torch.from_numpy(pred['bbox_abs']).float().unsqueeze(0)
                current_iou = box_iou(gt_bbox_tensor, pred_bbox_tensor)
                if current_iou > best_iou:
                    best_iou = current_iou
                    best_pred_idx = pred_idx

        if best_iou > iou_threshold:
            true_positives.append({'gt_box': gt, 'pred_box': preds_abs[best_pred_idx]})
            gt_boxes_abs[gt_idx]['matched'] = True
            preds_abs[best_pred_idx]['matched'] = True
        else:
            false_negatives.append(gt)

    # Any remaining unmatched predictions are False Positives
    for pred in preds_abs:
        if not pred['matched']:
            false_positives.append(pred)

    classified_detections[img_path] = {
        'TP': true_positives,
        'FP': false_positives,
        'FN': false_negatives,
        'image_dims': (img_width, img_height)
    }

print(f"Classified detections (TP, FP, FN) for {len(classified_detections)} images.")

total_initial_tp = 0
total_initial_fp = 0
total_initial_fn = 0

for img_path, detection_data in classified_detections.items():
    total_initial_tp += len(detection_data.get('TP', []))
    total_initial_fp += len(detection_data.get('FP', []))
    total_initial_fn += len(detection_data.get('FN', []))

print(f"\n--- Overall Initial TP, FP, FN Counts (before reclassification) ---")
print(f"Total True Positives: {total_initial_tp}")
print(f"Total False Positives: {total_initial_fp}")
print(f"Total False Negatives: {total_initial_fn}")



# Calculate overall metrics
overall_precision_initial = total_initial_tp / (total_initial_tp + total_initial_fp) if (total_initial_tp + total_initial_fp) > 0 else 0.0
overall_recall_initial = total_initial_tp / (total_initial_tp + total_initial_fn) if (total_initial_tp + total_initial_fn) > 0 else 0.0
overall_f1_score_initial = 2 * (overall_precision_initial * overall_recall_initial) / (overall_precision_initial + overall_recall_initial) if (overall_precision_initial + overall_recall_initial) > 0 else 0.0

print(f"\n--- Overall Metrics (before reclassification) ---")
print(f"Precision: {overall_precision_initial:.4f}")
print(f"Recall: {overall_recall_initial:.4f}")
print(f"F1 Score: {overall_f1_score_initial:.4f}")


# Calculate class-wise metrics
class_initial_tp = {i: 0 for i in range(len(class_names))}
class_initial_fp = {i: 0 for i in range(len(class_names))}
class_initial_fn = {i: 0 for i in range(len(class_names))}

for img_path, detection_data in classified_detections.items():
    for tp in detection_data.get('TP', []):
        class_initial_tp[tp['gt_box']['class_id']] += 1
    for fp in detection_data.get('FP', []):
        class_initial_fp[fp['class_id']] += 1
    for fn in detection_data.get('FN', []):
        class_initial_fn[fn['class_id']] += 1

print(f"\n--- Class-wise Metrics (before reclassification) ---")
for class_id, class_name in enumerate(class_names):
    tp = class_initial_tp[class_id]
    fp = class_initial_fp[class_id]
    fn = class_initial_fn[class_id]

    precision = tp / (tp + fp) if (tp + fp) > 0 else 0.0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0.0
    f1_score = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0.0

    print(f"\nClass: {class_name} (ID: {class_id})")
    print(f"  True Positives: {tp}")
    print(f"  False Positives: {fp}")
    print(f"  False Negatives: {fn}")
    print(f"  Precision: {precision:.4f}")
    print(f"  Recall: {recall:.4f}")
    print(f"  F1 Score: {f1_score:.4f}")

**Comment by Y-T-G** [comment]
You can see the FP, FN, TP boxes on the images by running validation with `visualize=True`

**Comment by beenajambucha1-design** [comment]
@Y-T-G But it only shows FP, FN, TP boxes with which are set in confusion matrix. As per model output, my actual FP, FN, and TP should be quite lesser than that.

**Comment by Y-T-G** [comment]
Thr confusion matrix is calculated using confidence threshold of 0.25 and IoU threshold of 0.45. You also need to use the same threshold.
